### PR TITLE
Convenience changes, tuple region macro

### PR DESCRIPTION
**Pull request by antiguru**
This PR adds convenience implementations for `ColumnStack` and:
* Macro to generate the tuple regions
* More types
* destructured copy for tuples.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -214,6 +227,34 @@ impl<T: Columnation> ColumnStack<T> {
...
+    /// not all elements are available as owned data.
+    ///
+    /// The element can be read by indexing
+    pub fn copy_destructured(&mut self, r0: &A, r1: &B, r2: &C) {
+        unsafe {
+            self.local.push(self.inner.copy_destructured(r0, r1, r2));
+        }
+    }
+}
+
```
Comment (frankmcsherry) [inline src/lib.rs]: Nit, but was this an example of a thing that could be generalized with `paste` as well?

File: src/lib.rs [hunk 2]
```diff
@@ -237,6 +278,53 @@ impl<T: Columnation> Default for ColumnStack<T> {
...
+impl<A: Columnation, B: Borrow<A>> std::iter::FromIterator<B> for ColumnStack<A> {
+    fn from_iter<T: IntoIterator<Item=B>>(iter: T) -> Self {
+        let iter = iter.into_iter();
+        let mut c = ColumnStack::<A>::with_capacity(iter.size_hint().0);
+        c.extend(iter);
+        c
+    }
+}
```
Comment (frankmcsherry) [inline src/lib.rs]: Nit, but elsewhere `T` is the `Columnation` type. Not always, now looking at it, but often.

File: src/lib.rs [hunk 3]
```diff
@@ -175,6 +177,17 @@ pub struct ColumnStack<T: Columnation> {
...
+    ///
+    /// Note that the associated region is not initialized to a specific capacity
+    /// because we can't generally know how much space would be required.
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            local: Vec::with_capacity(capacity),
+            inner: T::InnerRegion::default(),
+        }
+    }
+
```
Comment (frankmcsherry) [inline src/lib.rs]: I think we discussed this, and IIRC one take was that regions could have a `fn with_capacity_for(&[T]) -> Self` or something analogous, maybe an iterator, which would enable a better `clone_from` implementation. If we concluded that, no need to fix here but it might be worth leaving a note (IIRC, this method mainly exists to be used by another PR elsewhere, I think?).

**Review by frankmcsherry — APPROVED** [review]
These look good to me. A few comments that I wanted to sort out at some point, but needn't block merging!

File: src/lib.rs (hunk 1 above)
Comment (antiguru) [reply src/lib.rs]: Converted to macro.

File: src/lib.rs (hunk 2 above)
Comment (antiguru) [reply src/lib.rs]: Fixed, and removed the `Borrow` requirement. It turns out it's difficult to use this way.

File: src/lib.rs (hunk 3 above)
Comment (antiguru) [reply src/lib.rs]: The function is private now, because it's useful internally but doesn't do what a caller would expect it to do. A `fn with_capacity_for(&[T]) -> Self` seems useful, but it would require support of the regions, which we currently don't have. I'll leave this for later.

**Comment by antiguru** [comment]
I addressed your comments and opened #7 to capture the `with_capacity_for` idea. Thanks a bunch!

**Comment by frankmcsherry** [comment]
Thanks very much!
